### DISCUSSION TITLE
true() in XPath predicates

### DISCUSSION THREAD

**Discussion by timathom** [Jinntec/Fore#150] [General]
I am able to use the `true()` function in MIP XPath expressions, but it doesn't seem to work in control `@ref` predicates.

The example below throws an error when I try to use `true()` in a predicate ([log output](https://github.com/Jinntec/Fore/files/8993422/localhost-1656344737234.log)).

```
fontoxpath.esm.js:278 Uncaught (in promise) Error: XPST0017: Function Q{http://www.w3.org/2002/xforms}true with arity of 0 not registered. Did you mean "Q{http://www.w3.org/2005/xpath-functions}true ()"?
```

```
<fx-fore>
  <fx-model>
    <!-- Instance data -->
    <fx-instance>
      <data>
        <response>
          <one yes="true">This is true.</one>
          <two yes="false">This is false.</two>
        </response>
      </data>
    </fx-instance>
    <!-- /Instance data -->
  </fx-model>
  
  <!-- This works. -->
  <fx-output ref="response/*[@yes = 'true']"></fx-output>
  
  <!-- This doesn't. -->
  <fx-output ref="response/*[@yes = true()]"></fx-output>
  <fx-inspector open="true"></fx-inspector>
</fx-fore>
```

**Comment by timathom** [comment]
I seem to be getting the same error with `not()`, which is more problematic:

```
Uncaught (in promise) Error: XPST0017: Function Q{http://www.w3.org/2002/xforms}not with arity of 1 not registered. Did you mean "Q{http://www.w3.org/2005/xpath-functions}not (item()*)"?
```

**Reply by JoernT** [reply]
well - `response/*[@yes = true()]` is not true as it compares a boolean to a string - so that's correct. 

There's a boolean-from-string function for such cases that is adopted from XForms. 

However of course it shoudn't throw an error.

**Reply by timathom** [reply]
In BaseX,

```
let $test := <response yes="true"/>
return $test/@yes = true()
```

returns `true`.

[Same result](https://xpath.playground.fontoxml.com/?state=H4sIAIHwumIAAzXKuwqAIBgG0FeRbyoIbI6KXqG9pcs%2FBKnhJRTx3dPB7QwnQqiLMPQdvl3f%2B%2FGQwYCY0MGLJ3PUZF4lDbFAZtpgtaMNfC5hdaRDPrwevuTEJlZS0%2BZyKmnJ2xqRfiLqkYNxAAAA) in Fonto.

**Reply by JoernT** [reply]
hm, okay - that's evidence ;) Needs more investigation.

**Comment by DrRataplan** [comment]
Hm, this might be because the namespace overriding is not kicking in. In `xpath-evaluation.js`, there's this part: https://github.com/Jinntec/Fore/blob/dev/src/xpath-evaluation.js#L274 which tries to align the different prefixes. It tries to make `xforms` and `fn` fucntions available w/o a prefix. This must be going wrong here...

**Comment by JoernT** [comment]
just found in my use case that using 'fn:' prefix did the trick for me. 

The `globallyDeclaredFunctionLocalNames` array is empty for some reason.

So this part is failing:
```
            if (prefix === '' && globallyDeclaredFunctionLocalNames.includes(localName)) {
                // The function has been declared without a prefix and is called here without a prefix.
                // Just make this work. It is the developer-friendly way
                return {namespaceURI: 'http://www.w3.org/2005/xquery-local-functions', localName};
            }

```

**Comment by JoernT** [comment]
I'll convert this discussion to an issue.

**Comment by JoernT** [comment]
# workaround

Use 'fn:' prefix for core XPath functions

**Comment by DrRataplan** [comment]
I think I found it: for these functions, the correct functionNameResolver wasn't passed. See #155 for the fix!

**Reply by JoernT** [reply]
so its fixed and i mark this as answered
